Adds explosion helper proc w/ diminishing returns

nullbear opened the pull request:
Adds a new helper proc for calling explosions, which uses a 'power' integer instead of a
set of ranges.

Power (1 - 100) is based on explosion sizes (0, 0, 1) to (5, 10, 20) (the current maxcap setting.)
You can use a power of higher than 100 for larger explosions as well.

The formula i'm using makes explosions have diminishing returns on power, making larger grenades harder to make, but still possible. It should be used for keeping player-made bombs balanced within a sane range. but not used with hard-range coded bombs like c4.

Using current explosion math, the station can be reliably blown up (devastation range of 500) using 1000 units of explosion. (or 2000 nitroglycerin). 

Using my formula, it would require about 5 trillion units of water/potassium, or 500 million units of 'nitroglycerin'. So needless to say, it's effectively impossible to do in-game, even with max_cap disabled. 

`// ---------------------------------------------------------------------------------------------- //`

I've done some balance testing using maximum-size grenades (600u):
A **(2,5,10)** [link](http://i.imgur.com/Gn2tmYc.png?1)  can be made using water/potassium. (a quarter of the size it is now)
A **(3,7,14)** can be made using black powder or meth.
A **(6, 12, 24)** which is only slightly larger than a maxcap can be made with nitroglycerin.

Demonstration of new 'curve' using water/potass:
**200u** water/potass = 1, 3, 6 // large beaker grenade || Up from current
**600u** water/potass = 2, 5, 10 // halfcap bluespace beaker grenade || Down from current
**1800u** water/potass = 6, 13, 26 // maxcap chembomb payload (requires t4 parts)
**1000000u** water/potass = 10, 20, 40 // double maxcap, not possible to make.

Compare to old 'curve':
**200u** water/potass = 2, 5, 10 // large beaker grenade
**600u** water/potass = 6, 15, 30 // bluespace beaker grenade
**1800u** water/potass = 22, 45, 90 // maxcap chembomb payload
**1000000u** water/potass = 25000, 50000, 100000 // The Server Slayer

Do note, that the ranges are the radius of the explosion. And the actual hole typically affects about (x+1)^2 turfs for each unit of range. (ie a maxcap looks more like (25, 50, 100), and (22, 45, 90) as produced by chembomb payload is (484, 2025, 8100)

You can [****_CHECK THIS OUT**_**](https://www.desmos.com/calculator/mjzwoxpr9u) to see how the explosive power affects range. (Purple is old formula)
T is 'explosion power' and P is a multiplier. Multipliers for explosive reagents are: water/potass =
1/10, black powder/meth = 1/6, nitroglycerin = 1/2.

Replacing other explosions with this proc is not handled in this PR.

nullbear committed:
Adds new explosion proc

Adds a new proc for calling explosions, which uses 'power' instead of a
set of ranges.

Power (1 - 100) is based on minimum explosions size (0, 0, 1) to maximum
explosion size (5, 10, 20)

Formula keeps explosions within a 'sane' range. So while it is possible
to bypass maxcap using this, it will require a LOT of work.

I've done some balance testing, and a halfcap (2,5,10) can be made using
a double bluespace beaker of water and potassium. A (3,7,14) can be made
using black powder, and the largest bluespace grenade explosion can be
made using nitroglycerine, at (6, 12, 24) which is only slightly larger
than maxcap.

You can check out https://www.desmos.com/calculator/pgfz0gqqxf to see
how the explosive power affects range. T is 'units of explosion' and P
is a multiplier. Multipliers for explosive reagents are: water/potass =
1/10, black powder/meth = 1/6, nitroglycerin = 1/2.

kevinz000 commented:
no

nullbear commented:
y

francinum commented:
This is perfectly fine.

AnturK commented:
This is just a helper function, no need to freak out. That could use a better name.

PKPenguin321 commented:
So wait, new bomb formula that makes maxcaps impossible?

Jordie0608 commented:
No, read the PR description =|

MrStonedOne commented:
Change the proc name and the pr name and description

This shouldn't be marketed as a "new" explosion, but rather as a helper to translate a finite number to an explosion range.

nullbear commented:
> **PKPenguin321:** So wait, new bomb formula that makes maxcaps impossible?

Nah, it scales explosions linearly instead of exponentially, making bigger explosions exponentially harder to make (instead of your bomb either being 'shit' or 'maxcap' it's more likely to be something in between now.). Bluespace beaker maxcaps are no longer possible with "water/potassium". That's about it.

nullbear commented:
AH! That's the word for it. 

Diminishing returns! You get diminishing returns the bigger the explosion is. So maxcaps are still possible, just more expensive. And you can bypass it but it's harder as heck

PKPenguin321 commented:
>Bluespace beaker maxcaps are no longer possible with "water/potassium".

That's extremely lame

nullbear commented:
Use better chemicals.

PKPenguin321 commented:
The idea of better chems isn't that they're strictly better, it's that you need less of them for a bigger explosion (therefore not needing as-big beakers). Bluespace beaker water potassium maxcaps is fine

MoralineJester commented:
then buff water/potassium in general. This formula makes 'making' stronger bombs less of a waste of time.

nullbear commented:
I mean... i could buff potassium? But then we'd need a new cheap low-explosive chem to replace it.

MrStonedOne commented:
so, one issue is when admins change the max cap for certain memes. setting it to 1024 to disable it shouldn't basically turn simple chem grenades into server slayers

nullbear commented:
I'm doing a testmerge rn with maxcap removed.

Currently if you remove maxcap, you get server slayers. With this, even without maxcap chemistry explosions shouldn't be too crazy. The largest i was able to create using admin-shittery was about 15,30,60

nullbear commented:
I think i'm going to buff most chems by ~75%, to keep the bsbeaker balance about the same as before.

nullbear commented:
> **DEAD: Drake Langston says,** "chem nades are pretty gud right now"

Or maybe just 25%-50% at most.

MrStonedOne commented:
>I'm doing a testmerge rn with maxcap removed. ......

ok, so you still need to work on how you explain things a bit better. I was at first under the impression that it scaled to whatever the current bomb cap is, meaning if you raised the bomb cap, all explosions would become much much bigger.

This isn't the case, meaning you need to specify this some how.

nullbear commented:
It doesnt scale with maxcap whatsoever. But is still limited by maxcap if set to do so.

It offers diminishing returns on the explosive power of the mixture, which acts as its own natural limiter.

For example: 

To go from 1 range to 2 range, take 1 unit. to go from 2 range to 3, takes 5 units. To go from 3 to 4, takes 25 units, and so on... (not exact numbers, but it shows the diminishing returns)

This means that at higher levels, where explosive power:devastationrange is 100:5, to raise the range by one, it would require an additional 500 units, making it 600:6, and then again would be 3000:7, etc. 

This makes smaller bombs a lot easier to make. But larger ones are increasingly more expensive and difficult to make, typically limited by volume and reagents. (maximum volume of an chemical explosive is 1800, with a maximum explosive potential of 450 i think.) which would translate to a 6,12,24 explosion, which is only one range larger than maxcap. 

So with maxcap disabled, this still has an effective maxcap of 6,12,24 unless new reagents/volume containers are added. And still, that maxcap can only be reached by a player who worked HARD to achieve it.

------

Comparison table:

OLD | NEW
10 power = 50% knockdown | dented tiles.
50 power = minor brute damage | small breach.
75 power = dented tiles. | moderate breach
100 power = small breach | large breach
250 power = maxcap | maxcap
500 power = destroys half the station | tiny bit larger than maxcap
1000 power = destroys full station | small bit larger than maxcap

This is assuming maxcap is disabled. 
The new formula uses a curve which makes explosions much more useful, and you'll notice a difference as a bomber with how many units you use. Compared to current logic, which is essentially 'all or nothing', where either you get a pathetic tile denting, and add 50 more units and suddenly its a maxcap.

Because of the diminishing returns, it's also impossible to mass-produce large bombs, but possible to produce several 'medium' bombs, or many 'small' bombs.

With old logic, you can make several large bombs, but it's about just as expensive to make small bombs which aren't useful at all.

I greatly recommend you use the sliders on the desmos calculator links i've placed at the bottom of the PR body, to see for yourself better how the diminishing returns affect range and cost.

Fox-McCloud commented:
Annoying as hell proc, to say the least, and I"ll tell you why.

It operates under the assumption that the max cap will be 5-10-20. If this ever changes (higher or lower), then the calculations for this are going to be off (and I know there's downstreams that use different caps). "Power" level doesn't really tell you much about how devastation the explosion is, at a glance.

Something like 1-3-5 instantly tells me "well, it's going to destroy the turf it's on, has medium diameter of 5 and light radius of 9....where was "power 15" is nebulous, at best. You can't determine how "good" or "bad" that is at a glance other than knowing that 15 is inherently better than 14...but it's not clear *how much* better, as opposed to dealing with the raw numbers.

And yes, I know you can hypothetically compare it, to the max cap (ie: power 50 would be 3-5-10), but this is still far less readable and far less useful, in the long run.

So for that I give this a big :-1:

nullbear commented:
@Fox-McCloud 

This doesnt replace the (0,0,0) format of explosions, explicitly for that reason. And actually has a proc for easily getting an idea of the range that a particular power level provides. 

The procs i included are for being able to check the range of a given power, or alternatively, to get the 'power' of a given range. The only thing i could do better is provide a table of power:range for commonly used ranges.

Lastly, this is intended for use by scaleable explosions, the type that are made by players. And NOT hard-coded/static explosions, which would make better use out of the (0,0,0) format, for the added customization it provides.

Explosion (D,H,L) Used by static or custom explosions. Like c4, and minibombs.
Explosion (power) Used by dynamic explosions, to make use of diminishing returns, and use as a balancing point. (if you want a maxcap to require 1000 units of X, make each unit create 0.1 power. If you want a maxcap to require 10000 units of Y, make y create 0.01 power per unit.) It's not intended for static explosions, but rather for balancing dynamic explosions around 5,10,20 (a really big hole)

If you would like, i can add a scaling feature, which scales with a BOMBSCALE define which would be useful to servers which want bombs to be easier to make.

nullbear committed:
Clarity Changes

Renames dynamic explosion proc for clarity.

Makes dyn_x explosions scaleable with DYN_EX_SCALE variable.

Adds debug dyn_x verbs, and a dyn_x drop_bomb proc to fun admin verbs.

MrStonedOne commented:
>Power (1 - 100) is based on explosion sizes (0, 0, 1) to maxcap (5, 10, 20).

This line right here is whats confusing me, change it to

>Power (1 - 100) is based on explosion sizes (0, 0, 1) to (5, 10, 20) (current maxcap).

Because right now it reads like if I change mapcap from 20 to 40, 100 would equal 10, 20, 40

nullbear committed:
Adjusts formula

better efficiency and readability.

Adds debug proc for changing explosion scale.

Adds a comment containing common power/explosion sizes.

nullbear commented:
The scale isn't tied into maxcap. But i've added a scaling feature for use by admins/downstream.

The formula change does nothing (its the same formula, just written differently and without calling sqrt() )

nullbear committed:
oops

Bawhoppen commented:
walls of text reee

ChangelingRain commented on code/game/objects/explosion.dm:
```diff
@@ -246,4 +246,18 @@
+	range = round((2 * power)**DYN_EX_SCALE)
+	explosion(epicenter, round(range * 0.25), round(range * 0.5), round(range), flash_range*range, adminlog, ignorecap, flame_range*range, silent, smoke)
+
+// Using default dyn_ex scale:
```
This should note that it was 0.5 at the time of writing just in case ~someone~ changes it and doesn't update this.

nullbear replied:
its just the 

heavy = 1/4, mid = 1/2, light = 1  that's used everywhere else though.

optimumtact commented:
This was test merged yes? how did it go?

nullbear commented:
The testmerge went well. Syndie ship came and lots of grenades were made by CMO + chemist to blow it up. They blew it up, and said that the bombs were neither too strong nor unreasonably weak. So good.
